File: binarySearch/binarySearch.js
const nums = [
  -2,
  3,
  4,
  7,
  8,
  9,
  11,
  13,
  15,
  17,
  18,
  19,
  20,
  21,
  23,
  24,
  25,
  28,
  30
]

//===========================================================
// LINEAR SEARCH
//===========================================================
/* Assuming that the input array is sorted and has no duplicates */

const linearSearch = (arr, target) => {
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === target) {
      return i
    }
  }
  return -1
}

console.time('Liner: ')
console.log('linear', linearSearch(nums, 19))
console.timeEnd('Liner: ')

//===========================================================
// BINARY SEARCH
//===========================================================
/* Assuming that the input array is sorted and has no duplicates */

// Recursive
const binarySearchRecursive = (
  arr,
  target,
  left = 0,
  right = arr.length - 1
) => {
  if (left > right) return -1
  const mid = Math.floor(left + (right - left) / 2)

  if (arr[mid] === target) {
    return mid
  }
  if (target < arr[mid]) {
    return binarySearchRecursive(arr, target, left, mid - 1)
  }
  return binarySearchRecursive(arr, target, mid + 1, right)
}

console.time('BS Recursion: ')
console.log('recursive', binarySearchRecursive(nums, 19))
console.timeEnd('BS Recursion: ')

// Iterative
const binarySearchIterative = (arr, target) => {
  let left = 0
  let right = arr.length - 1

  while (left <= right) {
    const mid = Math.floor(left + (right - left) / 2)

    if (arr[mid] === target) {
      return mid
    }
    if (target < arr[mid]) {
      right = mid - 1
    } else {
      left = mid + 1
    }
  }
  return -1
}

console.time('BS Iteration: ')
console.log('iterative', binarySearchIterative(nums, 19))
console.timeEnd('BS Iteration: ')

File: python/binary_search.py
from math import floor
# Assume list is sorted and only contains numbers
# Assume that target can be not found in list

# [1,3,4,5,6,8,9]


def binary_search(list, target: int):
    midpoint = floor(len(list) / 2)
    left = 0
    right = len(list) - 1

    try:

        if target == list[left]:
            return left
        if target == list[right]:
            return right
        if target == list[midpoint]:
            return midpoint

        if target >= list[midpoint]:
            left = floor(len(list) / 2)

            while list[left] < target:
                left += 1

            if list[left] == target:
                return left

        elif target <= list[midpoint]:
            right = floor(len(list) / 2)

            while list[right] > target:
                right -= 1

            if list[right] == target:
                return right

    except IndexError:
        return -1


sample_list = [1, 3, 4, 5, 6, 8, 9]

print(binary_search(sample_list, 6))  # 4
print(binary_search(sample_list, 3))  # 1
print(binary_search(sample_list, 9))  # 6
print(binary_search(sample_list, 10))  # -1
